Lib/asyncio/locks.py

```diff
@@ -412,3 +412,218 @@ def release(self):
+        """Optionally run the 'action' and release the tasks waiting
+        in the barrier.
+        """
+        try:
+            if self._action:
+                await self._action()
+            # enter draining state
+            self._set_draining()
+            self._cond.notify_all()
+        except:
+            # an exception occurs during the _action coroutine,
+            # or the last calling task cancels 
+            # Break and reraise
+            self._break()
+            raise
+
+    async def _wait(self):
+        """Wait in the barrier until we are released.  Raise an exception
+        if the barrier is reset or broken.
+        """
+        # wait for end of filling 
+        # unless a CancelledError occurs
+        await self._cond.wait_for(lambda: not self._is_filling())
+
+        if self._is_broken() or self._is_resetting():
+            raise BrokenBarrierError
+
+    def _exit(self):
+        """If we are the last tasks to exit the barrier, signal any tasks
+        waiting for the barrier to drain.
+        """
+        if self._count == 0:
+            if self._is_resetting() or self._is_draining():
+                self._set_filling()                
+            self._cond.notify_all()
+
+    async def reset(self):
+        """Reset the barrier to the initial state.
+        Any tasks currently waiting will get the BrokenBarrier exception
+        raised.
+        """
+        async with self._cond:
+            if self._count > 0:
+                if not self._is_resetting():# self._is_filling() 
+                                            # or self._is_draining() 
+                                            # or self.is_broken()
+                    #reset the barrier, waking up tasks
+                    self._set_resetting()                
+            else:
+                self._set_filling()
+            self._cond.notify_all()
+
+    async def abort(self):
+        """Place the barrier into a 'broken' state.
+        Useful in case of error.  Any currently waiting tasks and tasks
+        attempting to 'wait()' will have BrokenBarrierError raised.
+        """
+        async with self._cond:
+            self._break()
+
+    def _break(self):
+        # An internal error was detected.  The barrier is set to
+        # a broken state all parties awakened.
+        self._set_broken()
+        self._cond.notify_all()
+
+    @property
+    def parties(self):
+        """Return the number of tasks required to trip the barrier."""
+        return self._parties
+
+    @property
+    def n_waiting(self):
+        """Return the number of tasks currently waiting at the barrier."""
+        if self._is_filling():
+            return self._count
+        return 0
+
+    @property
+    def n_blocking(self):
+        """Return the number of tasks currently blocking at the barrier."""
+        if self._is_draining():
+            return self._count_block
+        return 0
+
+    def _set_broken(self):
```

Comment:
There is a misunderstanding here.
I thought you keep `parties`, `n_waiting`, and `broken` attributes (they exist in `threading.Barrier` already).

All other attrs (including `n_waiting`) should not be exposed unless you can describe and protect an opposite opinion.

Reply:
> All other attrs (including n_waiting) should not be exposed unless you can describe and protect an opposite opinion.

Are you talking about `n_blocking`attr instead of `n_waiting` attr? I will remove it and restore `broken` attr.

> The cancellation behavior sounds reasonable.
> Please describe it in docs, the behavior is not very obvious.

Yes, I will do

> Sorry if I did not express my intention clearly.

It is ok for me, don't worry, your comments are very relevant

Reply:
Documentation is updated with cancellation task behavior.
`broken`attr is restored, `n_blocking` attr is removed.
Please @asvetlov could you approve running workflows. Thank